File: src/main/java/com/ocbc/auctionservice/services/UserService.java
package com.ocbc.auctionservice.services;

import com.ocbc.auctionservice.entities.User;
import com.ocbc.auctionservice.exceptions.users.UserAlreadyExistException;
import com.ocbc.auctionservice.exceptions.users.UserNotFoundException;
import com.ocbc.auctionservice.repositories.UserRepository;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Propagation;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@Slf4j
public class UserService {

    @Autowired
    private UserRepository userRepository;

    public List<User> getUsers() {
        log.info("Retrieve all users from database");
        return userRepository.findAll();
    }

    @Transactional(readOnly = true)
    public User getUser(int id) {
        return userRepository.findById(id).orElseThrow(
                () -> new UserNotFoundException(String.format("User's id %s does not exist", id)));
    }

    @Transactional(readOnly = true)
    public User getUser(String name) {
        return userRepository.findByName(name).orElseThrow(
                () -> new UserNotFoundException(String.format("User's name %s does not exist", name)));
    }

    @Transactional(readOnly = true)
    public User getUserByUsername(String username) {
        return userRepository.findByUsername(username).orElseThrow(
                () -> new UserNotFoundException(String.format("Username %s does not exist", username)));
    }

    @Transactional(propagation = Propagation.REQUIRES_NEW)
    public User createUser(User user) {
        return userRepository.save(user);
    }

    //Just a sample
//    public Account doSomethingWithAccount(Account account){
//        // This one refer to jira ticket 5506 https://jira.com/ticket/5506
//        if (account.someSortOfAccountCondition()){
//            //TODO: Setting this like this for now, going to change when we get more updates from BA
//            //TODO: Please remove this part of the code when BA comes back something
//        }
//    }

    @Transactional(propagation = Propagation.REQUIRED)
    public User updateUser(User user, int id) {
        User existingUser = getUser(id);
        existingUser.setAge(user.getAge());
        existingUser.setAddress(user.getAddress());
        existingUser.setRoles(user.getRoles());
        existingUser.setPermissions(user.getPermissions());
        return userRepository.save(existingUser);
    }

    public void deleteUser(int id) {
        User existingUser = getUser(id);
        userRepository.delete(existingUser);
    }
}

File: src/main/kotlin/com/auction/service/services/UserService.kt
package com.auction.service.services

import com.auction.service.entities.User
import com.auction.service.exceptions.UserAlreadyExistException
import com.auction.service.exceptions.UserNotFoundException
import com.auction.service.repositories.UserRepository
import lombok.extern.slf4j.Slf4j
import mu.KotlinLogging
import org.springframework.stereotype.Service

@Service
class UserService(
        private val userRepository: UserRepository
) {
    private val logger = KotlinLogging.logger {}

    fun getUsers(): List<User> {
        logger.info("Retrieving all users")
        return userRepository.findAll()
    }

    fun getUser(username: String): User {
        logger.info("Retrieving user $username")
        return userRepository.findByUsername(username) ?: throw UserNotFoundException("Unable to find user $username")
    }

//    fun findUser(id: Number) :  User{
//        return userRepository.findById(id).orElseGet(null) ?: throw Exception()
//    }

    fun getUser(id: Int): User {
        logger.info("Retrieving user $id")
        val optionalUser = userRepository.findById(id)
        if (optionalUser.isEmpty) throw UserNotFoundException("Unable to find user $id")
        return optionalUser.get()
    }

    fun createUser(user: User): User {
        logger.info("Create user ${user.name}")
        val existingUser = userRepository.findByUsername(user.username)
        if (existingUser != null) throw UserAlreadyExistException("Unable to create user ${user.name} due to username ${user.username} has already been taken.")
        return userRepository.save(user)
    }

    fun updateUser(newUser: User, id: Int): User {
        logger.info("Updating user ${newUser.name}")
        val existingUser = getUser(id)
        existingUser.apply {
            name = newUser.name
            username = newUser.username
            address = newUser.address
            password = newUser.password
            age = newUser.age
            roles = newUser.roles
        }
        return userRepository.save(existingUser)
    }

    fun deleteUser(id: Int) {
        logger.info("Deleting user $id")
        if (!userRepository.existsById(id)) throw UserNotFoundException("Unable to find user $id")
        userRepository.deleteById(id)
    }
}